format/format.go

```diff
@@ -0,0 +1,406 @@
+			if err := file.Release(releaseCtx); err != nil {
+				return fmt.Errorf("failed to release file: %w", err)
+			}
+		}
+
+		return nil
+	})
+}
+
+// match filters the file against global excludes and returns a list of formatters that want to process the file.
+func (c *CompositeFormatter) match(file *walk.File) []*Formatter {
+	// first check if this file has been globally excluded
+	if pathMatches(file.RelPath, c.globalExcludes) {
+		log.Debugf("path matched global excludes: %s", file.RelPath)
+
+		return nil
+	}
+
+	// a list of formatters that match this file
+	var matches []*Formatter
+
+	// otherwise, check if any formatters are interested in it
+	for _, formatter := range c.formatters {
+		if formatter.Wants(file) {
+			matches = append(matches, formatter)
+		}
+	}
+
+	return matches
+}
+
+// Apply applies the configured formatters to the given files.
+func (c *CompositeFormatter) Apply(ctx context.Context, files []*walk.File) error {
+	var toRelease []*walk.File
+
+	for _, file := range files {
+		matches := c.match(file) // match the file against the formatters
+
+		// check if there were no matches
+		if len(matches) == 0 {
+			// log that there was no match, exiting with an error if the unmatched level was set to fatal
+			if c.unmatchedLevel == log.FatalLevel {
+				return fmt.Errorf("no formatter for path: %s", file.RelPath)
+			}
+
+			log.Logf(c.unmatchedLevel, "no formatter for path: %s", file.RelPath)
+
+			// no further processing to be done, append to the release list
+			toRelease = append(toRelease, file)
+
+			// continue to the next file
+			continue
+		}
+
+		// record there was a match
+		c.stats.Add(stats.Matched, 1)
+
+		// check if the file is new or has changed when compared to the cache entry
+		if file.Cache == nil || file.Cache.HasChanged(file.Info) {
+			// add this file to a batch and if it's full, apply formatters to the batch
+			if key, batch := c.batches.Append(file, matches); len(batch) == c.batchSize {
+				c.apply(ctx, newBatchKey(matches), batch)
+				// reset the batch
+				c.batches[key] = make([]*walk.File, 0, c.batchSize)
+			}
+		} else {
+			// no further processing to be done, append to the release list
+			toRelease = append(toRelease, file)
+		}
+	}
+
+	// release files that require no further processing
+	// we set noCache to true as there's no need to update the cache, since we skipped those files
+	releaseCtx := walk.SetNoCache(ctx, true)
+
+	for _, file := range toRelease {
+		if err := file.Release(releaseCtx); err != nil {
+			return fmt.Errorf("failed to release file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// BustCache compares the currently configured formatters with their respective entries in the db.
+// If a formatter was added, removed or modified, we clear any path entries from the cache, ensuring that all paths
+// get formatted with the most recent formatter set.
+func (c *CompositeFormatter) BustCache(db *bolt.DB) error {
```

Comment:
(this feedback is not blocking for this PR, you just got me thinking about an architectural thing)

I have not read this function carefully. I think I understand the problem we're trying to solve here, though, and it looks like this code is very similar to stuff that existed in the now removed `format.go` file.

I am a little surprised that this is how we're solving the problem. I'd have expected the formatters + their configuration to be a piece of the "cache key" we use to cache files. For example, rather than caching "I formatted foo.c when it was modified Wednesday and was 128 KiB", we'd cache "I formatted foo.c with tool1 (with tool1-config42), tool2 (with tool2-config32) when it was modified Wednesday and was 128 KiB". If we were to cache this way, then we wouldn't need any special code to detect formatter changes, a formatter configuration change would just naturally result in a cache miss.

This makes me wonder: do we ever expire entries from the treefmt cache? Or does it grow forever?

Reply:
I think I just ported this from treefmt1. 


I don't remember seriously looking at the approach. You point out some valid things which I'll move into an issue to follow up after this is merged. 

We could just hash all the formatters and their configs together. If that hash changes, we invalidate the cache. 

As for expiring paths, we don't currently do that. I imagine in practice the caches get invalidated naturally.before.it becomes too much of a problem, but it's something we should have a solution for.

Reply:
Created to follow up:

- https://github.com/numtide/treefmt/issues/455
- https://github.com/numtide/treefmt/issues/456